// user/index.js
var md5 = require('md5');

function makeCodeCookie(taille)
{
  var text = "";
  var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  for( var i=0; i < taille; i++ ){
      text += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return text;
}

var champComplet = function(req){
  return("sexe" in req.body
    && "lastname" in req.body
    && "firstname" in req.body
    && "mailadress" in req.body
    && "birthdate" in req.body
    && "password" in req.body
    && "passwordbis" in req.body);
}

var verifInscription = function(req,res,next){
  var pg = require('pg');
  pg.defaults.ssl = true;
  pg.connect("postgres://iqohxmivsnsekz:83aeedca145e9d4c835c84cae8b2a3145416435e78c3c5ce82152b4c3e57c6d7@ec2-54-83-205-71.compute-1.amazonaws.com:5432/ddmteemcfd95nl"
  ,function(err, client) {
    if (err) throw err;
    client
      .query("SELECT * FROM public.user where mailadress = $1;",[req.body.mailadress])
      .on("end", function(result){
        if(result.rowCount == 0){
          if(champComplet(req))
          {
            next();
          }
          else {
            res.send({error: 'Des champs sont vides.'});
          }
        }
        else{
          res.send({error: 'Cette adresse mail est déjà utilisée.'});
        }
      });
  });
}

var inscription = function(req, res){
    var pg = require('pg');
    pg.defaults.ssl = true;
    var newUser = req.body;
    newUser.cookiecode = makeCodeCookie(30);
    pg.connect("postgres://iqohxmivsnsekz:83aeedca145e9d4c835c84cae8b2a3145416435e78c3c5ce82152b4c3e57c6d7@ec2-54-83-205-71.compute-1.amazonaws.com:5432/ddmteemcfd95nl"
    ,function(err, client) {
      if (err) throw err;
      client
        .query("INSERT into public.user (lastname,firstname,mailadress,password,registration,lastconnection,cookiecode,sexe,birthdate,stayconnected)values($1,$2,$3,$4,NOW(),NOW(),$5,$6,$7,false);"
        ,[newUser.lastname, newUser.firstname, newUser.mailadress, md5(newUser.password), newUser.cookiecode, newUser.sexe, newUser.birthdate])
        .on("end", function(result){
          res.send(newUser);
        });
    });
}

var verifConnection = function(req, res, next){
  var pg = require('pg');
  pg.defaults.ssl = true;
  pg.connect("postgres://iqohxmivsnsekz:83aeedca145e9d4c835c84cae8b2a3145416435e78c3c5ce82152b4c3e57c6d7@ec2-54-83-205-71.compute-1.amazonaws.com:5432/ddmteemcfd95nl"
  ,function(err, client) {
    if (err) throw err;
    client
      .query("SELECT password FROM public.user where mailadress = $1;",[req.body.mailadress])
      .on("end", function(result){
        if(result.rowCount != 0 && result.rows[0].password == md5(req.body.password)){
          next();
        }
        else {
          res.send({error: 'Mauvais identifiants'});
        }
      });
  });
}

var addStayConnected = function(req, res, next){
  var pg = require('pg');
  pg.defaults.ssl = true;
  var stayconnected;
  if(req.body.stayconnected){
    stayconnected = true;
  }
  else {
    stayconnected = false;
  }
  pg.connect("postgres://iqohxmivsnsekz:83aeedca145e9d4c835c84cae8b2a3145416435e78c3c5ce82152b4c3e57c6d7@ec2-54-83-205-71.compute-1.amazonaws.com:5432/ddmteemcfd95nl"
  ,function(err, client) {
    if (err) throw err;
    client
      .query("UPDATE public.user SET stayconnected = $1 WHERE mailadress = $2;",[stayconnected, req.body.mailadress])
      .on("end", function(result){
        next();
      });
  });
}

var connection = function(req, res, next){
    var pg = require('pg');
    pg.defaults.ssl = true;
    pg.connect("postgres://iqohxmivsnsekz:83aeedca145e9d4c835c84cae8b2a3145416435e78c3c5ce82152b4c3e57c6d7@ec2-54-83-205-71.compute-1.amazonaws.com:5432/ddmteemcfd95nl"
    ,function(err, client) {
      if (err) throw err;
      client
        .query("SELECT * FROM public.user where mailadress = $1;",[req.body.mailadress])
        .on('row', function(row) {
          res.send(row);
        });
    });
}

var connectionCookie = function(req, res){
  var pg = require('pg');
  pg.defaults.ssl = true;
  var resSend = false;
  pg.connect("postgres://iqohxmivsnsekz:83aeedca145e9d4c835c84cae8b2a3145416435e78c3c5ce82152b4c3e57c6d7@ec2-54-83-205-71.compute-1.amazonaws.com:5432/ddmteemcfd95nl"
  ,function(err, client) {
    if (err) throw err;
    client
      .query("SELECT * FROM public.user where cookiecode = $1;",[req.body.cookiecode])
      .on('end', function(result) {
        if(result.rowCount == 1){
          res.send(result.rows[0]);
        }
        else {
          res.send({error: 'Probleme de connexion.'});
        }
      });
  });
}

var verifPassword = function(req, res, next){
  var pg = require('pg');
  pg.defaults.ssl = true;
  pg.connect("postgres://iqohxmivsnsekz:83aeedca145e9d4c835c84cae8b2a3145416435e78c3c5ce82152b4c3e57c6d7@ec2-54-83-205-71.compute-1.amazonaws.com:5432/ddmteemcfd95nl"
  ,function(err, client) {
    if (err) throw err;
    client
      .query("SELECT password FROM public.user where id_user = $1;",[req.body.id_user])
      .on("end", function(result){
        console.log(result.rowCount + " : " + result.rows[0].password + " : " + (req.body.actualpassword));
        if(result.rowCount != 0 && result.rows[0].password == md5(req.body.actualpassword)){

          next();
        }
        else {
          res.send({error: 'Mauvais mot de passe.'});
        }
      });
  });
}

var editPassword = function(req, res){
  var pg = require('pg');
  pg.defaults.ssl = true;
  pg.connect("postgres://iqohxmivsnsekz:83aeedca145e9d4c835c84cae8b2a3145416435e78c3c5ce82152b4c3e57c6d7@ec2-54-83-205-71.compute-1.amazonaws.com:5432/ddmteemcfd95nl"
  ,function(err, client) {
    if (err) throw err;
    client
      .query("UPDATE public.user SET password = $1 WHERE id_user = $2;"
      ,[md5(req.body.newpassword), req.body.id_user])
      .on("end", function(result){
        res.send({});
      });
  });
}


exports.editPassword = editPassword;
exports.verifPassword = verifPassword;
exports.addStayConnected = addStayConnected;
exports.verifConnection = verifConnection;
exports.connection = connection;
exports.inscription = inscription;
exports.verifInscription = verifInscription;
exports.connectionCookie = connectionCookie;
